### PR TITLE
[Docs] Add the Byte-Pair Encoding Docs.

### PR DESCRIPTION
**Pull request by keli-wen**
## Description

- [x] Add the Byte-Pair Encoding Docs.

Refer to the https://github.com/Mixture-AI/meta-llama-explain/issues/17

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the `lint.sh` to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] Code is well-documented.
- [x] Related issue is referred in this PR.

### CODE REVIEW

File: docs/CN/BPE/README.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
+# Byte-Pair Encoding
+
+⏰ Read : `30min`
+
+Tokenize 的原理我认为是 NLP/LLM 领域的 Meta Knowledge 之一。而 BPE 可以说是当前最流行的分词算法，今天我结合一些前辈优秀的工作探索一下 BPE 算法的原理。
```
Comment (NANWOOD) [inline docs/CN/BPE/README.md:5]: Tokenize or Tokenizer?

File: docs/CN/BPE/README.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+- 为什么需要专门的分词算法，又为什么选择 BPE 而不是其他的压缩编码算法呢，例如 Huffman 编码？
+
+在目前的 NLP 范式中，输入通常是一段文本序列。在任何领域中的研究中，我们倾向于先确定最小单元，比如原子，普朗克时间。尽管后续的研究可能推翻这一结论，但确定最小单元有助于研究的开展。在 NLP 领域中，我们把 token 作为文本/序列数据中的最小离散单元。
+
+所以分词，便是将输入序列拆分为一组 token 的“观测”手段。它也和我们的日常息息相关，例如中文的**字词**，英文中以空格分割的**单词**等。所以一个很直觉的发问是：为什么不直接以空格作为分隔符进行分词呢？例如：`你好世界。 -> [你, 好, 世, 界, 。]` 或者 `I wen to New York last week. -> [i, went, to, new, york, last, week, .]`。
+
+我个人给出一些可能的原因：
+
+- 首先，直接按空格进行分词会**导致词表大小不必要的庞大**，如果希望限定词表大小还需要进一步考虑词表中 token 优先级的设定（即当容量填满时如何淘汰 token）
+- 其次，按空格分词会导致分词粒度太大。这会导致大量的 out of vocabulary （OOV） 的问题，并且没有**泛化性**。下面我们具体解释下泛化性。
```
Comment (NANWOOD) [inline docs/CN/BPE/README.md:22]:
```suggestion
- 其次，按空格分词会导致分词粒度太大。这会导致大量的 out of vocabulary （OOV） 的问题，并且缺乏**泛化性**。下面我们具体解释下泛化性。
```

File: docs/CN/BPE/README.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+例如按空格分词的词表为 `[rich, old, older, oldest, hard, harder, hardest]` ，而如果使用**子词** `er`，`est` ，词表变为 `[rich, old, hard, er, est]`，即节省了空间，还能将学到的知识迁移，避免一部分 OOV 问题（例如对于 `richest` 也可以进行合理的编码，而在原始的方案中会变成 `unknown`）
```
Comment (NANWOOD) [inline docs/CN/BPE/README.md:24]: “知识迁移” 有点抽象？

File: docs/CN/BPE/README.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+所以，用 sub-word （子词）进行编码的好处呼之欲出。既能节省空间，又能泛化解决 OOV 问题。
```
Comment (NANWOOD) [inline docs/CN/BPE/README.md:26]:
```suggestion
所以，用 sub-word （子词）进行编码的好处呼之欲出。既能节省空间，又能提升泛化性来解决 OOV 问题。
```

File: docs/CN/BPE/README.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+> 为什么选择 BPE 而不是其他的压缩编码算法呢，例如 Huffman 编码？
+
+这个问题可以参考论文中的解释。，
+
+*“--- The main difference to other compression algorithms, such as Huffman encoding, which have been proposed to produce a variable-length encoding of words for NMT (Chitnis and DeNero, 2015), is that our symbol sequences are still interpretable as subword units, and that the network can generalize to translate and produce new words (unseen at training time) on the basis of these subword units.”* **《Neural Machine Translation of Rare Words with Subword Units》Section2.2**
+
+主要区别在于，BPE 生产的**符号序列**仍然可以理解和解释为子词单位，并且模型可以基于这些子词单位推广，去理解那些没有见过（不在词表中）的新词，解决 OOV 问题。
+
+其中，符号序列我认为可以立即为词表中的 token。对于 Huffman 编码，生成的词表可能是 `101`，`010` 的比特序列，相比于 BPE 生成的类似于 `[low, er, est]`，后者可以更好的进行泛化。
```
Comment (NANWOOD) [inline docs/CN/BPE/README.md:38]:
```suggestion
其中，符号序列我认为可以理解为词表中的 token。对于 Huffman 编码，生成的词表可能是 `101`，`010` 的比特序列，相比于 BPE 生成的类似于 `[low, er, est]`，后者可以更好的进行泛化。
```

File: docs/CN/BPE/README.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+所以总的来说，BPE 的优势是什么呢？
+
+- 简单，基于贪心的算法思想使得其实现更简洁。
+- 可以避免词表不必要庞大的问题，并且其范式更科学（可以轻松的指定期望的词表大小）
+- 可以很好的处理 OOV 问题（Out of Vocabulary），且避免陷入 Character Embedding 方法导致分词粒度太细的问题。而后文中提到的字节级 BPE 实现可以直接避免 OOV 问题。
```
Comment (NANWOOD) [inline docs/CN/BPE/README.md:44]:
```suggestion
- 可以很好的处理 OOV 问题（Out of Vocabulary），且避免陷入 Character Embedding 方法所导致的分词粒度太细的问题。而后文中提到的字节级 BPE 实现可以直接避免 OOV 问题。
```

File: docs/CN/BPE/README.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+> - `Normlization` 可以是 **统一转化为小写** 等类似的操作。例如，`Hello World! -> hello wolrd!`
+> - `Pre-tokenization` 一般是按空格拆分输入序列。例如，`hello world! -> [hello, world, !]`
+>
+> 想要了解更多可以参考 Reference 中的 [Hugging Face Learn: Normalization and pre-tokenization](https://huggingface.co/learn/nlp-course/chapter6/4)。
+
+```
+"hug", "pug", "pun", "bun", "hugs"
+```
+
+**Step1 基础词表** 首先，我们会有一个基础词典。`["b", "g", "h", "n", "p", "s", "u"]`。通常情况下，我们应该至少保证基础词点包括了所有的 ASCII 码字符。甚至一部分 Unicode 字符。如果你需要 tokenize 的序列存在训练集中没出现过的字符，那么就会导致 unknown token 的出现（也就是 OOV 问题）
```
Comment (NANWOOD) [inline docs/CN/BPE/README.md:67]:
```suggestion
**Step1 基础词表** 首先，我们会有一个基础词典。`["b", "g", "h", "n", "p", "s", "u"]`。通常情况下，我们应该至少保证基础词典包括了所有的 ASCII 码字符，甚至一部分 Unicode 字符。如果需要 tokenize 的序列存在训练集中没出现过的字符，那么就会导致 unknown token 的出现（也就是 OOV 问题）。
```

File: docs/CN/BPE/README.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+*“--- The main difference to other compression algorithms, such as Huffman encoding, which have been proposed to produce a variable-length encoding of words for NMT (Chitnis and DeNero, 2015), is that our symbol sequences are still interpretable as subword units, and that the network can generalize to translate and produce new words (unseen at training time) on the basis of these subword units.”* **《Neural Machine Translation of Rare Words with Subword Units》Section2.2**
+
+主要区别在于，BPE 生产的**符号序列**仍然可以理解和解释为子词单位，并且模型可以基于这些子词单位推广，去理解那些没有见过（不在词表中）的新词，解决 OOV 问题。
+
+其中，符号序列我认为可以立即为词表中的 token。对于 Huffman 编码，生成的词表可能是 `101`，`010` 的比特序列，相比于 BPE 生成的类似于 `[low, er, est]`，后者可以更好的进行泛化。
+
+所以总的来说，BPE 的优势是什么呢？
+
+- 简单，基于贪心的算法思想使得其实现更简洁。
+- 可以避免词表不必要庞大的问题，并且其范式更科学（可以轻松的指定期望的词表大小）
```
Comment (NANWOOD) [inline docs/CN/BPE/README.md:43]:
```suggestion
- 可以避免词表不必要庞大的问题，并且其范式更科学（可以轻松的指定期望的词表大小）。
```

File: docs/CN/BPE/README.md (hunk 2 above)
Comment (keli-wen) [reply docs/CN/BPE/README.md:22]: Done

File: docs/CN/BPE/README.md (hunk 4 above)
Comment (keli-wen) [reply docs/CN/BPE/README.md:26]: Done

File: docs/CN/BPE/README.md (hunk 6 above)
Comment (keli-wen) [reply docs/CN/BPE/README.md:44]: Get

File: docs/CN/BPE/README.md (hunk 8 above)
Comment (keli-wen) [reply docs/CN/BPE/README.md:43]: Done

**Review by NANWOOD — APPROVED** [review]
LGTM.

**Pull request merged**